packages/client-sdk-nodejs/src/config/middleware/middleware.ts

```diff
@@ -10,10 +12,19 @@ export interface MiddlewareRequestHandler {
   onResponseStatus(status: StatusObject): Promise<StatusObject>;
 }
 
+export interface MiddlewareRequestHandlerContext {
+  [key: string]: string | number | boolean;
+}
+
 /**
  * The Middleware interface allows the Configuration to provide a higher-order function that wraps all requests.
  * This allows future support for things like client-side metrics or other diagnostics helpers.
+ *
+ * An optional context can be provided that is essentially a <key, value> map {@link MiddlewareRequestHandlerContext}.
+ * The context object is available to each individual invocation of the request handler for the middleware.
  */
 export interface Middleware {
-  onNewRequest(): MiddlewareRequestHandler;
+  onNewRequest(
+    context?: MiddlewareRequestHandlerContext
+  ): MiddlewareRequestHandler;
```

Comment:
and i agree with you that it could be a nice side effect for users, though I don't think we should expose it to them until we have a use case.

Reply:
Yes that was the first thing I verified and you're right otherwise the build would've failed for the other middlewares as they still implement it. Also the advanced example that I ran using these changes uses one middleware with this and one without so that also served as an e2e test.

Reply:
Also to convince myself i had checked that if I remove the `?` which means it's not optional; the other middlewares started complaining

Reply:
Well, TIL:

"In TypeScript, a class is considered to implement an interface as long as the class has the required methods with the correct return types. The parameters of the class's methods are allowed to be a subset of the parameters defined in the interface, meaning they can be less restrictive. This is based on TypeScript's structural subtyping."

So even if I don't make this optional, it'd still work and TS will consider the classes to be implementing the interface! That's a LOL moment for me coming primarily from Java but they might have had good reasons.